File: apps/android/domain/src/main/kotlin/com/sanvya/app/domain/dashboard/Trend.kt
package com.sanvya.app.domain.dashboard

import java.time.LocalDate

/** How far back the Expense-trends tile looks. Web's `TrendPeriod`. */
enum class TrendPeriod(val key: String) {
    THREE_DAYS("3d"),
    ONE_WEEK("1w"),
    ONE_MONTH("1m"),
    ONE_YEAR("1y");

    companion object {
        fun from(key: String?): TrendPeriod = entries.firstOrNull { it.key == key } ?: ONE_MONTH
    }
}

/**
 * One bucket: the date it starts on, and what was spent in it.
 *
 * The START DATE, not a label. Web's `buildTrend` returns `"12 Aug"` built from
 * a hardcoded English month array; formatting a date is the view's job on a
 * platform that has a locale, and returning the label here would have shipped
 * English into two otherwise localised apps.
 */
data class TrendBucket(val startIso: String, val totalMinor: Long)

/**
 * Buckets daily expense totals into period-appropriate buckets.
 *
 * A port of web's `buildTrend`, with `today` passed in rather than read from the
 * clock — which is what makes it testable, and is why there are vectors for it.
 *
 * The bucket shapes are web's, exactly:
 * - 3d / 1w: one bucket per day, oldest first.
 * - 1m: FOUR buckets of seven days each, and web's windows **overlap by a day**
 *   (`start = today - (w*7 + 6)`, `end = today - w*7`, inclusive both ends). A
 *   day that is a boundary is counted in two buckets. That is web's arithmetic
 *   and it is preserved deliberately: correcting it here alone would make the
 *   same month read differently in the browser and on the phone.
 * - 1y: twelve calendar months, oldest first.
 *
 * @param dailyTotals `YYYY-MM-DD` to minor units, as the query returns them.
 */
fun buildTrend(
    dailyTotals: Map<String, Long>,
    period: TrendPeriod,
    todayIso: String,
): List<TrendBucket> {
    // An unparseable `today` is a caller bug, not something to crash a tile
    // over: Swift's `buildTrend` guards its date parse and returns empty, and a
    // golden vector pins the pair. `LocalDate.parse` throws where Swift returns
    // nil, so the difference has to be spelled out here or the two platforms
    // disagree on the same input -- which is exactly what the vector caught.
    val today = runCatching { LocalDate.parse(todayIso) }.getOrNull() ?: return emptyList()
    return when (period) {
        TrendPeriod.THREE_DAYS, TrendPeriod.ONE_WEEK -> {
            val n = if (period == TrendPeriod.THREE_DAYS) 3 else 7
            (n - 1 downTo 0).map { i ->
                val day = today.minusDays(i.toLong())
                TrendBucket(day.toString(), dailyTotals[day.toString()] ?: 0L)
            }
        }
        TrendPeriod.ONE_MONTH -> (3 downTo 0).map { w ->
            val start = today.minusDays((w * 7 + 6).toLong())
            val end = today.minusDays((w * 7).toLong())
            // String comparison, not LocalDate: the keys are zero-padded ISO
            // dates, so lexical order IS chronological order and there is
            // nothing to parse per row.
            val startKey = start.toString()
            val endKey = end.toString()
            val sum = dailyTotals.entries.sumOf { (key, value) ->
                if (key >= startKey && key <= endKey) value else 0L
            }
            TrendBucket(start.toString(), sum)
        }
        TrendPeriod.ONE_YEAR -> (11 downTo 0).map { m ->
            val first = today.withDayOfMonth(1).minusMonths(m.toLong())
            val prefix = "%04d-%02d".format(first.year, first.monthValue)
            val sum = dailyTotals.entries.sumOf { (key, value) ->
                if (key.startsWith(prefix)) value else 0L
            }
            TrendBucket(first.toString(), sum)
        }
    }
}

/** One month of the cashflow series. */
data class CashflowMonth(
    /** `YYYY-MM`. */
    val month: String,
    val incomeMinor: Long,
    val expenseMinor: Long,
) {
    val netMinor: Long get() = incomeMinor - expenseMinor
}

/**
 * Folds `(yearMonth, type, total)` rows into one entry per month.
 *
 * Web's `useCashflow`, including the trailing-eight window. Rows arrive sorted
 * by month from the query, and the fold preserves that order rather than
 * re-sorting, so a month with only income and a month with only expense stay
 * where the query put them.
 */
fun monthlyCashflow(rows: List<Triple<String, String, Long>>, months: Int = 8): List<CashflowMonth> {
    val byMonth = LinkedHashMap<String, CashflowMonth>()
    for ((yearMonth, type, total) in rows) {
        val current = byMonth[yearMonth] ?: CashflowMonth(yearMonth, 0, 0)
        byMonth[yearMonth] = when (type) {
            "income" -> current.copy(incomeMinor = total)
            else -> current.copy(expenseMinor = total)
        }
    }
    return byMonth.values.toList().takeLast(months)
}

File: apps/ios/Domain/Sources/Domain/Trend.swift
import Foundation

/// How far back the Expense-trends tile looks. Web's `TrendPeriod`.
public enum TrendPeriod: String, CaseIterable, Sendable {
    case threeDays = "3d"
    case oneWeek = "1w"
    case oneMonth = "1m"
    case oneYear = "1y"

    public static func from(_ key: String?) -> TrendPeriod {
        guard let key, let value = TrendPeriod(rawValue: key) else { return .oneMonth }
        return value
    }
}

/**
 One bucket: the date it starts on, and what was spent in it.

 The START DATE, not a label. Web's `buildTrend` returns `"12 Aug"` built from a
 hardcoded English month array; formatting a date is the view's job on a platform
 that has a locale, and returning the label here would have shipped English into
 two otherwise localised apps.
 */
public struct TrendBucket: Sendable, Equatable {
    public let startIso: String
    public let totalMinor: Int64

    public init(startIso: String, totalMinor: Int64) {
        self.startIso = startIso
        self.totalMinor = totalMinor
    }
}

/**
 Buckets daily expense totals into period-appropriate buckets.

 A port of web's `buildTrend`, with `today` passed in rather than read from the
 clock — which is what makes it testable, and is why there are vectors for it.

 The bucket shapes are web's, exactly:
 - 3d / 1w: one bucket per day, oldest first.
 - 1m: FOUR buckets of seven days each, and web's windows **overlap by a day**
   (`start = today - (w*7 + 6)`, `end = today - w*7`, inclusive both ends). A day
   that is a boundary is counted in two buckets. That is web's arithmetic and it
   is preserved deliberately: correcting it here alone would make the same month
   read differently in the browser and on the phone.
 - 1y: twelve calendar months, oldest first.

 - Parameter dailyTotals: `YYYY-MM-DD` to minor units, as the query returns them.
 */
public func buildTrend(
    _ dailyTotals: [String: Int64],
    period: TrendPeriod,
    todayIso: String
) -> [TrendBucket] {
    // UTC, and its own Calendar rather than one of Domain's existing Ymd
    // helpers: `parseYmd` here returns a 0-based-month FinanceYmd and the only
    // `addDays` is private inside Recurring.swift. Borrowing either would have
    // meant widening something for one caller.
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = TimeZone(secondsFromGMT: 0)!
    guard let today = trendDate(todayIso, calendar) else { return [] }

    func iso(_ date: Date) -> String { trendIso(date, calendar) }
    func shift(_ date: Date, days: Int) -> Date {
        calendar.date(byAdding: .day, value: days, to: date) ?? date
    }

    switch period {
    case .threeDays, .oneWeek:
        let n = period == .threeDays ? 3 : 7
        return (0..<n).reversed().map { i in
            let day = shift(today, days: -i)
            return TrendBucket(startIso: iso(day), totalMinor: dailyTotals[iso(day)] ?? 0)
        }
    case .oneMonth:
        return (0..<4).reversed().map { w in
            let start = shift(today, days: -(w * 7 + 6))
            let end = shift(today, days: -(w * 7))
            let startKey = iso(start)
            let endKey = iso(end)
            // String comparison, not Date: the keys are zero-padded ISO dates,
            // so lexical order IS chronological order and there is nothing to
            // parse per row.
            let sum = dailyTotals.reduce(Int64(0)) { acc, entry in
                (entry.key >= startKey && entry.key <= endKey) ? acc + entry.value : acc
            }
            return TrendBucket(startIso: startKey, totalMinor: sum)
        }
    case .oneYear:
        var firstOfThisMonth = calendar.dateComponents([.year, .month], from: today)
        firstOfThisMonth.day = 1
        let anchor = calendar.date(from: firstOfThisMonth) ?? today
        return (0..<12).reversed().map { m in
            let first = calendar.date(byAdding: .month, value: -m, to: anchor) ?? anchor
            let prefix = String(iso(first).prefix(7))
            let sum = dailyTotals.reduce(Int64(0)) { acc, entry in
                entry.key.hasPrefix(prefix) ? acc + entry.value : acc
            }
            return TrendBucket(startIso: iso(first), totalMinor: sum)
        }
    }
}

private func trendDate(_ iso: String, _ calendar: Calendar) -> Date? {
    let parts = String(iso.prefix(10)).split(separator: "-")
    guard parts.count == 3, let y = Int(parts[0]), let m = Int(parts[1]), let d = Int(parts[2]) else { return nil }
    var components = DateComponents()
    components.year = y; components.month = m; components.day = d
    return calendar.date(from: components)
}

private func trendIso(_ date: Date, _ calendar: Calendar) -> String {
    let c = calendar.dateComponents([.year, .month, .day], from: date)
    return String(format: "%04d-%02d-%02d", c.year ?? 0, c.month ?? 0, c.day ?? 0)
}

/// One month of the cashflow series.
public struct CashflowMonth: Sendable, Equatable {
    /// `YYYY-MM`.
    public let month: String
    public let incomeMinor: Int64
    public let expenseMinor: Int64

    public var netMinor: Int64 { incomeMinor - expenseMinor }

    public init(month: String, incomeMinor: Int64, expenseMinor: Int64) {
        self.month = month
        self.incomeMinor = incomeMinor
        self.expenseMinor = expenseMinor
    }
}

/**
 Folds `(yearMonth, type, total)` rows into one entry per month.

 Web's `useCashflow`, including the trailing-eight window. Rows arrive sorted by
 month from the query, and the fold preserves that order rather than re-sorting,
 so a month with only income and a month with only expense stay where the query
 put them.
 */
public func monthlyCashflow(_ rows: [(String, String, Int64)], months: Int = 8) -> [CashflowMonth] {
    var order: [String] = []
    var byMonth: [String: CashflowMonth] = [:]
    for (yearMonth, type, total) in rows {
        if byMonth[yearMonth] == nil {
            order.append(yearMonth)
            byMonth[yearMonth] = CashflowMonth(month: yearMonth, incomeMinor: 0, expenseMinor: 0)
        }
        let current = byMonth[yearMonth]!
        byMonth[yearMonth] = type == "income"
            ? CashflowMonth(month: yearMonth, incomeMinor: total, expenseMinor: current.expenseMinor)
            : CashflowMonth(month: yearMonth, incomeMinor: current.incomeMinor, expenseMinor: total)
    }
    let all = order.compactMap { byMonth[$0] }
    return all.count > months ? Array(all.suffix(months)) : all
}
